packages/runner/sync-customer-os-data/service/sync_service.go

```diff
@@ -111,151 +109,92 @@ func (s *syncService) syncEmailMessages(ctx context.Context, dataService common.
+				logrus.Errorf("failed merge interaction session with external reference %v for tenant %v :%v", message.ExternalId, tenant, err)
 			}
 
-			var fromContactId string
-
-			if message.Direction == entity.INBOUND {
-				fromContactId, err = s.repositories.ContactRepository.GetFirstOrCreateContactByEmail(
-					ctx, tenant, message.FromEmail, message.FromFirstName, message.FromLastName, message.ExternalSystem)
+			if !failedSync {
+				interactionEventId, err = s.repositories.InteractionEventRepository.MergeInteractionEvent(ctx, tenant, syncDate, message)
 				if err != nil {
 					failedSync = true
-					logrus.Errorf("failed creating contact with email %v for tenant %v :%v", message.FromEmail, tenant, err)
+					logrus.Errorf("failed merge interaction event with external reference %v for tenant %v :%v", message.ExternalId, tenant, err)
 				}
-			}
 
-			// set initiator for new conversation
-			if messageCount == 0 {
-				initiatorUsername = message.FromEmail
-
-				if message.Direction == entity.OUTBOUND {
-					initiator := entity.ConversationInitiator{
-						ExternalSystem: message.ExternalSystem,
-						ExternalId:     message.UserExternalId,
-						FirstName:      message.FromFirstName,
-						LastName:       message.FromLastName,
-						Email:          message.FromEmail,
-						InitiatorType:  entity.USER,
-					}
-					err := s.repositories.ConversationRepository.UserInitiateConversation(ctx, tenant, conversationId, initiator)
-					if err != nil {
-						failedSync = true
-						logrus.Errorf("failed set user initiator for conversation %v in tenant %v :%v", conversationId, tenant, err)
-					}
-				} else if message.Direction == entity.INBOUND {
-					initiator := entity.ConversationInitiator{
-						Id:            fromContactId,
-						FirstName:     message.FromFirstName,
-						LastName:      message.FromLastName,
-						Email:         message.FromEmail,
-						InitiatorType: entity.CONTACT,
-					}
-					err := s.repositories.ConversationRepository.ContactInitiateConversation(ctx, tenant, conversationId, initiator)
+				if !failedSync {
+					err = s.repositories.InteractionEventRepository.MergeInteractionEventToSession(ctx, tenant, interactionEventId, sessionId)
 					if err != nil {
 						failedSync = true
-						logrus.Errorf("failed set contact initiator for conversation %v in tenant %v :%v", conversationId, tenant, err)
+						logrus.Errorf("failed to associate interaction event to session %v for tenant %v :%v", message.ExternalId, tenant, err)
 					}
 				}
 			}
 
-			// set contact participants
-			if len(fromContactId) > 0 {
-				err := s.repositories.ConversationRepository.ContactByIdParticipateInConversation(ctx, tenant, conversationId, fromContactId)
+			//from
+			if message.Direction == entity.OUTBOUND && !failedSync {
+				err := s.repositories.InteractionEventRepository.InteractionEventSentByUser(ctx, tenant, interactionEventId, message.UserExternalId, message.ExternalSystem)
```

Comment:
for outbound emails there is also from email address.
Let's use this email address, validating before that it belongs to user.
If email address is missing, or not linked to user. we can create it / link to user creator and still use it. Like for inbound emails)